Mathlib/RingTheory/Algebraic/Integral.lean

```diff
@@ -146,15 +156,249 @@ end Field
+    Algebra.IsAlgebraic R A :=
+  ⟨fun _ ↦ (alg.1 _).restrictScalars_of_isIntegral _ inj⟩
+
+theorem IsAlgebraic.trans_isIntegral [Algebra.IsAlgebraic R S] [int : Algebra.IsIntegral S A] :
+    Algebra.IsAlgebraic R A :=
+  ⟨fun _ ↦ (int.1 _).trans_isAlgebraic _ inj⟩
+
+end Algebra
+
+variable (R S)
+/-- If `R` is a domain and `S` is an arbitrary `R`-algebra, then the elements of `S`
+that are algebraic over `R` form a subalgebra. -/
+def Subalgebra.algebraicClosure [IsDomain R] : Subalgebra R S where
+  carrier := {s | _root_.IsAlgebraic R s}
+  mul_mem' ha hb := ha.mul hb
+  add_mem' ha hb := ha.add hb
+  algebraMap_mem' := isAlgebraic_algebraMap
+
+theorem integralClosure_le_algebraicClosure [IsDomain R] :
+    integralClosure R S ≤ Subalgebra.algebraicClosure R S :=
+  fun _ ↦ IsIntegral.isAlgebraic
+
+theorem Subalgebra.algebraicClosure_eq_integralClosure {K} [Field K] [Algebra K S] :
+    algebraicClosure K S = integralClosure K S :=
+  SetLike.ext fun _ ↦ isAlgebraic_iff_isIntegral
+
+instance [IsDomain R] : Algebra.IsAlgebraic R (Subalgebra.algebraicClosure R S) :=
+  (Subalgebra.isAlgebraic_iff _).mp fun _ ↦ id
+
+variable {R S}
+
+theorem Algebra.isAlgebraic_adjoin_iff [IsDomain R] {s : Set S} :
+    (adjoin R s).IsAlgebraic ↔ ∀ x ∈ s, IsAlgebraic R x :=
+  Algebra.adjoin_le_iff (S := Subalgebra.algebraicClosure R S)
+
+theorem Algebra.isAlgebraic_adjoin_of_nonempty [NoZeroDivisors R] {s : Set S} (hs : s.Nonempty) :
+    (adjoin R s).IsAlgebraic ↔ ∀ x ∈ s, IsAlgebraic R x :=
+  ⟨fun h x hx ↦ h _ (subset_adjoin hx), fun h ↦
+    have ⟨x, hx⟩ := hs
+    have := (isDomain_iff_noZeroDivisors_and_nontrivial _).mpr ⟨‹_›, (h x hx).nontrivial⟩
+    isAlgebraic_adjoin_iff.mpr h⟩
+
+theorem Algebra.isAlgebraic_adjoin_singleton_iff [NoZeroDivisors R] {s : S} :
+    (adjoin R {s}).IsAlgebraic ↔ IsAlgebraic R s :=
+  (isAlgebraic_adjoin_of_nonempty <| Set.singleton_nonempty s).trans forall_eq
+
+theorem IsAlgebraic.of_mul [NoZeroDivisors R] {y z : S} (hy : y ∈ nonZeroDivisors S)
+    (alg_y : IsAlgebraic R y) (alg_yz : IsAlgebraic R (y * z)) : IsAlgebraic R z := by
+  have ⟨t, ht, r, hr, eq⟩ := alg_y.exists_nonzero_eq_adjoin_mul hy
+  have := alg_yz.mul (Algebra.isAlgebraic_adjoin_singleton_iff.mpr alg_y _ ht)
+  rw [mul_right_comm, eq, ← Algebra.smul_def] at this
+  exact this.of_smul (mem_nonZeroDivisors_of_ne_zero hr)
+
+namespace Transcendental
+
+section
+
+variable {a : A} (ha : Transcendental R a)
+include ha
+
+lemma extendScalars_of_isIntegral [NoZeroDivisors S] [Algebra.IsIntegral R S]
+    (inj : Function.Injective (algebraMap S A)) : Transcendental S a := by
+  contrapose ha
+  rw [Transcendental, not_not] at ha ⊢
+  exact ha.restrictScalars_of_isIntegral _ inj
+
+lemma extendScalars [NoZeroDivisors S] [Algebra.IsAlgebraic R S]
+    (inj : Function.Injective (algebraMap S A)) : Transcendental S a := by
+  contrapose ha
+  rw [Transcendental, not_not] at ha ⊢
+  exact ha.restrictScalars _ inj
```

Comment:
There is some evidence that the [NoZeroDivisors S] condition might be weakened. If S is a localization of R, then the injectivity of `R[X] → A` (from `Transcendental R a`) implies the injectivity of `S[X] → S ⊗[R] A`, but since A/S/R is a tower and S/R is a localization, `S ⊗[R] A` is just `A`, and we get `Transcendental S a` without requiring S to have no zero divisors. Of course, `Transcendental R a` implies `algebraMap R S` is injective, so the submonoid can't contain any zero divisors.